### DISCUSSION TITLE
Feature to generate a report of all battery-powered devices, including device name, device type, battery type, and last battery replaced date.

### DISCUSSION THREAD

**Discussion by 42Network** [andrew-codechimp/HA-Battery-Notes#419] [Ideas]
### Checklist

- [X] I have filled out the template to the best of my ability.
- [X] This only contains 1 feature request (if you have multiple feature requests, open one feature request for each feature request).
- [X] This issue is not a duplicate feature request of [previous feature requests](https://github.com/andrew-codechimp/ha-battery-notes/issues?q=is%3Aopen+is%3Aissue+label%3Aenhancement).

### Is your feature request related to a problem? Please describe.
Thank you for this awesome contribution to the HA ecosystem. I hope to see it included in the main branch eventually!

I would like to know my entire inventory of battery-powered devices and group by battery type to know how many spare batteries I should have on hand in the worst case. E.g. How many 2032, 2045, AA, AAA, etc. should I have on hand as spares, based on how long the batteries last while in service versus the shelf-life of new, unused batteries?

### Describe the solution you'd like

I would like to generate a report of all battery-powered devices, including device name, device type, battery type, and last battery replaced date.

### Describe alternatives you've considered

The Device list will show all the "Battery Notes" devices with name, type, and battery level (see screenshot below), but I would also like to see the battery type and install date.

### Additional context

![2024-01-08_10-36-52](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/35043496/6627d003-3c56-4e7a-82c8-2af19ade03b2)

**Comment by andrew-codechimp** [comment]
Thanks for the feedback, I'm not sure this will ever make it into core as code quality and time commitments wouldn't allow me to maintain the standards required for an official integration.

I like the idea of a "report" but as you're probably aware the Devices screen can't be customised so it would have to be a completely new Frontend component, a bit like Calendar and Alarmo amongst other add.  This is a very different development skill from integration development (HTML, TypeScript etc) rather than Python so it would be a whole new area for me to learn.

In the meantime myself and others are using autoentities to create custom dashboard items, not as feature rich but shows what batteries we need soon.  There's a good example here https://community.home-assistant.io/t/custom-component-battery-notes/613821/24

I'm closing this issue and moving it to the discussion/ideas area so it's not lost.

**Comment by ThomDietrich** [comment]
Hey guys,
I think this can be achieved via a custom Lovelace configuration.

I didn't have the time to dig much deeper so far. Currently I have a good overview of battery levels ONLY but it's technically feasible to add the battery type information. I didn't come up with the logic to link the two entities yet. Any ideas?

https://community.home-assistant.io/t/low-battery-level-detection-notification-for-all-battery-sensors/258664/345?u=thomdietrich

![c5d87bedfa4cdf5053d5dc9e6339d56747a8a838](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/2870104/3df58de3-e9d5-4a8e-8872-18d3172a9e51)

**Comment by andrew-codechimp** [comment]
There is an interesting custom card yaml config over on the HA forums that merges both battery level & type that may be useful

https://community.home-assistant.io/t/custom-component-battery-notes/613821/24?u=codechimp

**Reply by ThomDietrich** [reply]
That looks also great! I might post my combined version later.

**Reply by andrew-codechimp** [reply]
I'm going to be introducing a docs section with community examples of things like this soon so I'll copy and call out credit when I do.
I find the HA forums a bit tricky to navigate when people update examples.

**Reply by ThomDietrich** [reply]
That would be wonderful.

I found that all of these solutions out there work with entity ids and therefore do not work if your HA instance uses a non-english language.
What do you think, would it be possible and/or make sense to add the battery type as attribute to the battery level?

**Reply by andrew-codechimp** [reply]
The _battery_type suffix of the entity id should be english but there was a bug before a few versions ago that didn't do this.
I can't add an attribute to the battery level sensor as you can't update another platforms sensors which is annoying as that would have been really useful.

From 1.3.5 release note
> There was a bug in previous versions where battery note sensors were created with entity_id's in non-english for those with a HA system language of Slovak, German, Hungarian and Danish. New battery note entity_id's will be created in English as per HA guidelines but existing entities could be incorrect depending on when you added them and you should check/delete/recreate the battery notes for these. This will cause issues if you start using automations/cards based on entity_id. Entity names are still translated. Apologies for this issue.

**Reply by ThomDietrich** [reply]
Some integrations started to generate localized entity_ids as well. On my system I can see this combination atm:
 - `sensor.button_batterie` (ZHA)
 - `sensor.button_battery_type` (Battery Note)

Which makes bringing both together complicated, especially talking about a one-fits-all community solution to post in your community examples docs section.

> I can't add an attribute to the battery level sensor as you can't update another platforms sensors which is annoying as that would have been really useful.

Too bad! That would have been great indeed. However, you can add attributes to your integration freely. How about adding the battery level or the `source_entity` similarly to how powercalc does it?

![image](https://github.com/andrew-codechimp/HA-Battery-Notes/assets/2870104/80a01859-54df-4744-a2ff-53a69bad0154)

**Reply by andrew-codechimp** [reply]
Hmm, I thought I was the only one who'd messed up entity_id naming.
Adding a battery_percentage attribute to the type is a great idea, I'd not thought of flipping the direction! 

I am looking at listening to the battery level change for a different reason so I have a bit of prototype code already, once I have that working it will be easy to add an attribute and get the battery_type entity to change state in line with the battery percentage so you can base any automations off it as well.

I have a few smaller changes planned first but this will be the next big priority item I'll work on.

**Reply by ThomDietrich** [reply]
> battery_type entity to change state in line with the battery percentage

Based on experience, this is a great idea! Won't hurt the average user but e.g. notifications in the low battery level case will be a lot easier if we can just concentrate on the "supercharged" battery type entity ;)

While at it, I would suggest to add both battery_percentage and battery_level_entity_id as attributes. The latter could be very helpful to look up details of this entity in templating logic.

> Hmm, I thought I was the only one who'd messed up entity_id naming.

I'm not sure what the latest on this is. There was an effort to localize entity friendly names but I can't find a clear guideline on entity_ids. You? https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations (btw my personal preference is that entity_ids should be in English, while friendly names can be localized)

**Reply by andrew-codechimp** [reply]
Now your comment of "supercharged" got me thinking.  I should be able to create a new entity that echo's the battery for the device a bit like the "Change device type of switch" helper which gives you a drop in replacement battery entity but with all the additions of battery notes as attributes.  Should even be able to hide the current battery.  Then all these fancy automations/cards can just reference the device class of battery, and extend to look at the attributes.

The battery_level_entity_id would maybe become irrelevant then but I can always add it as that's the first step in getting any of this to work.

I've not seen anything definite on entity_id naming either, just the first comment from an admin here seems to say it shouldn't be translated.
https://community.home-assistant.io/t/multi-language-entity-id-how-aliases/31339

**Reply by ThomDietrich** [reply]
You know what, I would absolutely love that!!

Not too sure whether it is needed or a good idea to hide the original battery. I would avoid touching what isn't my business :) But I agree, the following sounds good to me:
An entity like `sensor.button_battery_status` (status, info, extended, plus, ...) with the state and all attributes of the original `sensor.button_battery`, plus `battery_type` and `battery_quantity`. The entity updates with every battery level change, also resolving your forced_update idea from above. Agree, battery_level_entity_id would not be needed anymore. I'm really looking forward to this!

Bonus thought: See my screenshot from above. Common question I have when a device is unavailable is: "Was the battery low before it went offline", versus any other reason for a device to be offline. Do you have a thought how we could use the supercharged battery entity to provide such an info?

**Reply by andrew-codechimp** [reply]
I will think more about the hiding idea, maybe an optional flag in configuration.yaml but I like the way the switch as x does that.

I'll also add the attribute for last_replaced to the super battery entity, one place to query everything.

For another feature I am going to be looking at reading history to determine past battery levels so this could be used for a "last known level" once I get round to working out how to do it.

**Reply by andrew-codechimp** [reply]
A sneak peak of what this will look like here https://github.com/andrew-codechimp/HA-Battery-Notes/discussions/570

**Reply by ThomDietrich** [reply]
Btw, I just added a new device via ZHA and entity_id are german'ized: For example `button.eckstehlampe_identifizieren` instead of `button.eckstehlampe_identify`

**Reply by andrew-codechimp** [reply]
The more I look at ZHA the more I'll be sticking with z2m. It will make sharing of automations, blueprints or yaml cards a nightmare. 
Entity ids shouldn't be localised, there's a dismissed feature request here for it. 

https://community.home-assistant.io/t/multi-language-entity-id-how-aliases/31339
